cloudinit/config/modules.py

```diff
@@ -35,28 +38,31 @@ def form_module_name(name):
     return canon_name
 
 
-def fixup_module(mod, def_freq=PER_INSTANCE):
-    if not hasattr(mod, "frequency"):
-        setattr(mod, "frequency", def_freq)
-    else:
-        freq = mod.frequency
-        if freq and freq not in FREQUENCIES:
-            LOG.warning("Module %s has an unknown frequency %s", mod, freq)
-    if not hasattr(mod, "distros"):
-        setattr(mod, "distros", [])
-    if not hasattr(mod, "osfamilies"):
-        setattr(mod, "osfamilies", [])
-    return mod
+def validate_module(mod, name):
+    if (
+        not hasattr(mod, "meta")
+        or "frequency" not in mod.meta
+        or "distros" not in mod.meta
+    ):
+        raise ValueError(
+            f"Module '{mod}' with name '{name}' MUST have a 'meta' attribute "
+            "of type 'MetaSchema'."
+        )
+    if mod.meta["frequency"] not in FREQUENCIES:
```

Comment:
+1 noisy errors in misconfigured modules instead of silent DEFAULT to PER_INSTANCE. Strict behavior for the win.